### PR TITLE
Update: Refactor paragraph block to be a functional component

### PR DESCRIPTION
**Pull request by jorgefilipecosta**
## Description
This PR updates the paragraph block to be a functional component.
We are introducing new hook based API's (e.g: the colors hook) and these API's can only be used if the component is functional.

## How has this been tested?
I verified the paragraph still works as before.

### CODE REVIEW

File: packages/block-library/src/paragraph/edit.js [hunk 1]
```diff
@@ -45,174 +45,215 @@ const applyFallbackStyles = withFallbackStyles( ( node, ownProps ) => {
...
+		},
+		[ attribute, setAttributes ]
+	);
+}
+
+function ParagraphToolbar( { direction, setDirection } ) {
+	const isRTL = useSelect( ( select ) => {
+		return select( 'core/block-editor' ).getSettings().isRTL;
+	} );
+	const toolbarControls = useMemo(
```
Comment (youknowriad) [inline packages/block-library/src/paragraph/edit.js]: In the previous version, we were not using a memoized prop, do you think it's really necessary here?  I always wonder if we're over-optimizing sometimes, My personal opinion would be to start with a non-optimized version unless we measure change in performance.

File: packages/block-library/src/paragraph/edit.js [hunk 2]
```diff
@@ -45,174 +45,215 @@ const applyFallbackStyles = withFallbackStyles( ( node, ownProps ) => {
...
+function ParagraphPanelColor( {
+	backgroundColor,
+	fallbackBackgroundColor,
+	fallbackTextColor,
+	fontSize,
+	setBackgroundColor,
+	setTextColor,
+	textColor,
+} ) {
+	const colorSettings = useMemo(
```
Comment (youknowriad) [inline packages/block-library/src/paragraph/edit.js]: Same as above

File: packages/block-library/src/paragraph/edit.js [hunk 3]
```diff
@@ -45,174 +45,215 @@ const applyFallbackStyles = withFallbackStyles( ( node, ownProps ) => {
...
+function useAttributeSetter( attribute, setAttributes ) {
+	return useCallback(
+		( newValue ) => {
+			setAttributes( { [ attribute ]: newValue } );
+		},
+		[ attribute, setAttributes ]
+	);
+}
+
+function ParagraphToolbar( { direction, setDirection } ) {
```
Comment (youknowriad) [inline packages/block-library/src/paragraph/edit.js]: Should this be renamed ParagraphRTLToolbar instead?

**Review by youknowriad — COMMENTED** [review]
I started by commenting on one `useMemo`/`useCallback` but I noticed that it's a trend across all props and callbacks. I do wonder if we should use these two hooks in our refactorings from component classes to hooks. I understand that sometimes it's necessary to avoid rerenders but I'd probably prefer to start without trying to optimize everything.

Other than that, the changes are good.

**Review by youknowriad — APPROVED** [review]
Looks great, thanks 👍

**Comment by aduth** [comment]
In investigating [typing performance regressions](https://make.wordpress.org/core/2019/11/27/whats-new-in-gutenberg-27-november/) between Gutenberg 6.9 and Gutenberg 7.0, something as significant a refactor to the paragraph block like this seems like it could be a prime candidate as a contributor. It would have been good to do some performance measurements of this refactor (though generally I'd have hoped only for improvements).

**Pull request merged**